governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java

```diff
@@ -133,33 +143,60 @@ public void run() {
 
     }
 
-    private Deque<Callable<Void>> cleanupActions = new ConcurrentLinkedDeque<>();
+    private Map<Object, UnscopedCleanupAction> cleanupActions = new WeakHashMap<>();
     private ScopeCleaner scopeCleaner = new ScopeCleaner();
-
-    Map<Class<? extends Annotation>, Scope> scopeBindings;
+    private Map<Class<? extends Annotation>, Scope> scopeBindings;
 
     public PreDestroyMonitor(Map<Class<? extends Annotation>, Scope> scopeBindings) {
         this.scopeBindings = new HashMap<>(scopeBindings);
     }
 
     public <T> boolean register(T destroyableInstance, Binding<T> binding, Iterable<LifecycleAction> action) {
-        return scopeCleaner.isRunning() ? binding.acceptScopingVisitor(
-                new ManagedInstanceScopingVisitor(destroyableInstance, binding.getSource(), action)) : false;
+        if (scopeCleaner.isRunning()) {
+            boolean visitNoScope = Optional
+                    .ofNullable(binding.acceptTargetVisitor(new DefaultBindingTargetVisitor<T, Boolean>() {
```

Comment:
How expensive is this operations?  Wondering if this should be moved into a cache.

Reply:
To improve readability can you extract this anonymous class into a separate class.

Reply:
I extracted the anonymous class. Don't think the visitor is expensive; all the info is static structure / simple method calls